### ISSUE TITLE
Clarification about registering singletons with more than one token

### ISSUE THREAD

**Issue by 85danf** [microsoft/tsyringe#99]
**Describe the bug**
Hi there! first of all, great library - it makes usage and testing really convenient so kudos to you :)

I'm not sure this actually qualifies as a bug but still, it seems to me like unexpected behavior, but I'll let you judge:

Suppose I have a class which handles the database connection pool, `DbHandle` which I define as follows:
```javascript
@singleton()
@injectable()
export default class DbHandle implements PostConstruct { 
   
 constructor(private readonly clusterApi?: ClusterApi) {
            this.clusterApi = clusterApi;
        }

   public async init() {
        ...
   }
...
```
Where my `PostConstruct` interface simply has an `init()` function and `ClusterApi` is another class decorated with `@singleton`

The motivation for this is that DB initialization, for instance, must be made async (which is not possible in a constructor). So I Register it at the bottom of the `DbHandle.ts` file like this:
```javascript 
container.register<PostConstruct>('PostConstruct', DbHandle);
```
which works great when in my `main()` function I call:
```javascript
const needInit: PostConstruct[] = container.resolveAll('PostConstruct');
await Promise.all(needInit.map(item => item.init()));
```
(I have another class needing this kind of init but lets keep it simple)

Up until here, everything works as expected, the dependency container resolves the token, constructs the class and its dependency and then the init function is called.

My issue is that now I reference `DbHandle` from a third class:
```javascript
@singleton()
@injectable()
export default class SomeBusinessLogic {

    constructor(private readonly dbHandle: DbHandle) {
        this.dbHandle = dbHandle;
    }
...
```
At this point in my `main()` I call:
```javascript
await container.resolve(SomeBusinessLogic).someBusinessfunction();
```
Which causes a second `DbHandle` class to get constructed and be used instead of the one I already registered with the `PostConstruct` token.

I guess using `container.registerInstance()` can be used here to mitigate but my issue with this solution is that my `DbHandle` class is marked `@injectable` itself, so I shouldn't call its constructor manually.

Another solution might be to `container.resolve()` all of these init-requiring classes in my `main()`  so I have the instance (which was correctly constructed by the container) and then register the instances with the `PostConstruct` token as well but this means my main will need knowledge of these classes so I could init them there and then just as well and that kind of beats the purpose...

I understand interfaces have no type information at runtime which is probably the cause for this behavior but I'm wondering if there's any recommended way to achieve what I want apart from what I suggested above?

Thanks!

**To Reproduce**

1. Register a singleton under some text token, (semantically unrelated to its actual functionality - in my case `"PostConstruct"`
2. Call a function by resolving the text token (interface) from the container - works as expected.
3. Try resolving the singleton using its class as a token 

**Expected behavior**

Expected: both `container.resolveAll('InterfaceName')` and `container.resolve(Class)` should produce the same instance since it is marked `@singleton`

Actual: the container holds 2 instances, one for each token.

**Version**:
4.0.1

**Comment by avin-kavish** [comment]
How would the container know that the string token and singleton decorator are referring to the same singleton? There's no type information at run-time in js, as far as the container can see they are two distinct registrations.

**Comment by 85danf** [comment]
Hi and thanks for replying! 
I get your point about type information (like I mentioned in my description above), but coming from Spring (Java) I'm used to (and agree with) singletons being, well, singletons...
The whole concept of constructing more them of them kinda beats the purpose IMO.

BTW in Spring you can register a bean with both a string identifier and an annotation (effectively resolving a bean by its interface - using `@Qualifier("")`) or even register 2 classes as beans under the same interface and default to one with another annotation (`@Primary`).

In any case since that's not applicable here, what I ended up doing was to:
1. register my singletons when they're constructed `container.registerInstance<PostConstruct>('PostConstruct', this);`
2. Resolve them through the container prior to my `init()` method so that they are constructed and the instances can now be resolved with the `'PostConstruct'` token.

This works of course but also makes me have to know about all of the classes that require an init somewhere instead of just having them declare themselves as so.

And back to my question: is there any way to solve this currently using tsyring?

**Comment by avin-kavish** [comment]
Hmm... you have a different question above the expected behavior, I was talking about why your expected behavior will most likely not happen.

> This works of course but also makes me have to know about all of the classes that require an init somewhere instead of just having them declare themselves as so.

From your description, not even spring seams to be able to solve this problem correct? Are you saying that when annotated, spring discovers and calls an (async) init method automatically?

The problem you are having and the one you are trying to solve seems different. The problem of registering with the IOC container with a decorator and a string token are separate from the problem of asynchronous lazy initialisation, is it not?

What is your exact use case for wanting to instantiate & initialise eagerly?

**Comment by 85danf** [comment]
Perhaps I wasn't explaining myself correctly, let me try again please:

My specific issue is that I have some init logic that must happen asynchronously, like initializing a database connection or retrieving an authentication token from a REST endpoint, therefore it cannot happen within the constructor.

In Spring I would solve this simply by adding a `@PostConstruct` decorator to a public method (they are called annotations in Java) which will cause spring to invoke it at some point after the construction of the class but always before the application's main entrypoint is invoked.
Async calls are inherently different in Java in the sense that I can choose to wait for them to happen within the invoking thread (while other threads are free to do other stuff) or just invoke them and not care about when they happen (thus 'freeing' the current thread) - this mechanism in Java is usually achieved using thread pools. 

It's also worth mentioning that the mode I used to work with spring was to have it construct (and thus call `@PostConstruct`) in an eager manner so once my application's main point starts, I know all of my classes (referred to as `Beans` in Spring) have been constructed and my `@PostConstruct` logic was invoked already.

IMHO this is the whole point of having an IoC container in the first place - It does all the heavy lifting of setting up my app's logistics and then it 'calls' my application to start.
With tsyringe it seems like the container is waiting for me to invoke it and lazily constructs classes (which is fine from a dependency injection point of view) but also makes it troublesome to invoke async calls in an automatic manner, so it seems I'm left with having to invoke the async init logic at some central point.

My second point was that in Spring, unlike tsyringe, you usually register _interfaces_ and not classes (since it uses proxy classes under the hood) - then you have a few options how to tell Spring how to 'locate' the implementing class: 
1. Simply implementing it if there's only one implementation - Spring scans for classes in the application's `classpath` (a Java concept which essentially means the filesystem path where all class files reside)
2. Using a `@Primary` decorator to tell it to always resolve a specific class (while you can reference other implementing classes by their actual types)
3. Using a `Qualifier("")` decorator which registers a class under an interface but also under a string qualifier which you can use in any dependent class' constructor later on to resolve a specific implementation.
In the application I used to work on we also had a custom decorator called `@Init` which we 'told' Spring to call at a certain point in the initialization process - but that's a specific feature which I don't really care about here since as I demonstrated I could 'almost' get the same behavior with tsyringe.


It's obvious to me that tsyringe was created with a certain usecase in mind and I totally respect that, and its also very clear the mechanisms I described above rely on type information that's an inherent part of the Java language which Typescipt does not have.

I chose to use tsyring since it was _way_ simpler to integrate compared to other IoC modules I saw, and that's a very positive thing, so this is not about telling you how you _should_ implement your library but more of an open question about how you would recommend I solve my usecase given the way tsyringe does ioC.

Thanks for your time and willingness to answer.

**Comment by avin-kavish** [comment]
Right, that's much more clear :) First of all, I am not a maintainer and I don't make the decisions. I guess a `@PostConstruct` decorator can be provided by the library.  (Maybe you should make one and give us feedback) I'm just commentating based on my experience and understanding of IoC and DI.

> but also makes it troublesome to invoke async calls in an automatic manner, so it seems I'm left with having to invoke the async init logic at some central point.

The `@PostConstruct` is handy but it's not the only approach to the async initialization problem. For example, in ASP.NET, databases and other network or asynchronous connections are not initialized till just before they are needed. Such a pattern would look like

```ts
export default class DbHandle {
  private  _clusterApi
  private _init: Promise<any>

   get clusterApi() {
      if (!this._clusterApi.isConnected) {
          this._init = this._clusterApi.connect()
      }
      return _clusterApi
   }

   async function query(...args) {
    const cluster = this.clusterApi
    await this._init
    cluster.query(...args)
  }
}
```

This is _true_ lazy initialization because technically, there is no saying whether the resource is needed or not just after construction. If your use case allows for it, I would recommend this way. But this has a bit of boilerplate associated with it.

The most TSyringy way to post-construct initialise I can think of is to use a post construct aware factory when registering your singletons.

```ts
import {instanceCachingFactory} from "tsyringe";

container.register(
  'DbHandleSingleton',
  { 
    useFactory: instanceCachingFactory<DbHandle>(c => { 
      const handle = c.resolve(DbHandle) 
      await handle.init()
      return handle
    })
  }
)
```

We can extract this into a reusable function as follows,
```ts
function instanceCachingPostConstructFactory<T>(token: InjectionToken<T>) {
  return instanceCachingFactory<T>(c: DependencyContainer) => {
    const instance = c.resolve(token) 
    await instance.init()
    return instance
  })
}
```

But once again the token limitation is real.

**Comment by MeltingMosaic** [comment]
The idea of a decorator that can be applied to a method that executes that method after resolution doesn't seem crazy to me, and I think we would accept PRs for it.

**Comment by avin-kavish** [comment]
Are we going to call it `@PostConstruct`? Perhaps,

-  `@Initialise`
- `@AfterResolve`
- `@PostInject`
- `@PostResolve`

**Comment by MeltingMosaic** [comment]
I thought about this a little more, and I think we only need a post-construction initializer right now - there may be the need for a post-resolution decorator in the future, and we should leave the door open to that, but I think the desire is just an initializer. As for naming, `@initialize` might work, but is a bit generic, so maybe `@afterConstruction` might be more specific.

**Comment by avin-kavish** [comment]
Honestly, I'm not sure if this is such a good idea. Because there is no guarantee that the objects will be initialized in time for when they are used since the initialization completes asynchronously after the object is constructed. I guess this is considered okay in Java land? But this would be a big no no in C# land. A better pattern is to lazily initialize asynchronous things like db connections when they are needed. For example, check and connect just in time for a DB query inside the query method itself. Something like,

```ts
class DbHandle {

 async init() {
   return this.clusterApi.connect()
 }

  async query(...args) {
     await this.init()
     return this.clusterApi.query(...args)
  }

}
```

This is how the node postgres connection pool works and this is how Entity Framework works.

The other safe method is asynchronous dependency resolvers. [InversifyJS supports it.](https://stackoverflow.com/questions/40880447/how-to-inject-an-asynchronous-dependency-in-inversify) Although, I don't believe this will work for parameter based synchronous injection.

**Comment by 85danf** [comment]
The scenario you describe would of course not be ok in Java land. 
The most primary commitment I need fulfilled, as you point out, is to know that when I use my injected dependency it was already constructed and inited.

Spring's way of doing this is by being giving you a framework in which your code is executed, thus the `Bean`'s lifecycle is managed by Spring and not by the app (although you have hooks where you can intervene).
That way you have total certainty of what gets executed when (simplistically: Bean construction -> Post Construction -> Bean ready).
It is true that you can reach situations in Spring where, due to order-of-initialization or circular dependencies, these promises are broken but that usually implies bad design and not an issue with the IoC framework itself.

I don't like the InversifyJS way you linked since it means that I (i.e. the user of a functionality) need to _know_ whether it is resolved async - so I would have to write something like 
```await getDb()``` which kind of beats the purpose of what I'm suggesting IMO.

The Spring way of letting you 'know' when everything is ready and you can start your app (by means of a `ContextReady` hook for example) is preferable since it removes ambiguity and provides a clear promise of dependency lifecycle.

**Comment by bohdan-shulha** [comment]
Hi there.

I have met the similar issue, however I can provide a bit different reproducible demo:

```typescript

import "reflect-metadata";
import { container } from "tsyringe";

const TOKEN = Symbol.for("token");

class ServiceProvider {
  constructor() {
    console.log("service initialized"); // will be printed two times
  }
}

container.registerSingleton(TOKEN, ServiceProvider);

console.log(container.resolve(TOKEN) === container.resolve(ServiceProvider));
```

In this case I am unsure whether `container.resolve(ServiceProvider)` should be successful at all as I have registered the service under specified token but a service available simply by using a class.

@Xapphire13 @MeltingMosaic  what do you think of this?

**Comment by bohdan-shulha** [comment]
Oh, just discovered these lines

```typescript

// No registration for this token, but since it's a constructor, return an instance
if (isConstructorToken(token)) {
  return this.construct(token, context);
}
```

So, this is intentionally and I understand why, but I simply didn't expect this behavior in this exact case.

IDK if we have to do anything with this.

For now, I just put the "guard" which prevents unwanted injections:

```typescript
di.registerSingleton(TOKEN, updateSource);
di.register(updateSource, {
  useFactory: () => {
    throw new Error(
      "Cannot inject TOKEN. Use @inject(TOKEN) instead",
    );
  },
});
```

**Comment by Xapphire13** [comment]
@85danf, sorry for such a delayed response. This problem can be solved by registering a class which takes care of the post initialization, as shown in the following passing test:

```typescript
  @singleton()
  class DbHandle {
    isInitialized = false;

    init() {
      this.isInitialized = true;
    }
  }

  @injectable()
  class PostConstruct {
    // Add other dependencies as needed
    constructor(dbHandle: DbHandle) {
      dbHandle.init();
    }
  }

  const handle = globalContainer.resolve(DbHandle);
  expect(handle.isInitialized).toBeFalsy();
  globalContainer.resolve(PostConstruct);
  expect(handle.isInitialized).toBeTruthy();
```

Feel free to re-open if this still doesnt solve your problem
